src/flow/src/adapter/flownode_impl.rs

```diff
@@ -13,24 +13,312 @@
+                    reason: format!("Invalid flow type: {}", flow_type),
+                }
+                .fail()
+            }
+        };
+
+        let flow_id = args.flow_id;
+        let src_table_ids = args.source_table_ids.clone();
+
+        let res = match flow_type {
+            FlowType::Batching => self.batching_engine.create_flow(args).await,
+            FlowType::Streaming => self.streaming_engine.create_flow(args).await,
+        }?;
+
+        self.src_table2flow
+            .write()
+            .unwrap()
+            .add_flow(flow_id, flow_type, src_table_ids);
+
+        Ok(res)
+    }
+
+    async fn remove_flow(&self, flow_id: FlowId) -> Result<(), Error> {
+        let flow_type = self.src_table2flow.read().unwrap().get_flow_type(flow_id);
+        match flow_type {
+            Some(FlowType::Batching) => self.batching_engine.remove_flow(flow_id).await,
+            Some(FlowType::Streaming) => self.streaming_engine.remove_flow(flow_id).await,
+            None => FlowNotFoundSnafu { id: flow_id }.fail(),
+        }?;
+        // remove mapping
+        self.src_table2flow.write().unwrap().remove_flow(flow_id);
+        Ok(())
+    }
+
+    async fn flush_flow(&self, flow_id: FlowId) -> Result<usize, Error> {
+        let flow_type = self.src_table2flow.read().unwrap().get_flow_type(flow_id);
+        match flow_type {
+            Some(FlowType::Batching) => self.batching_engine.flush_flow(flow_id).await,
+            Some(FlowType::Streaming) => self.streaming_engine.flush_flow(flow_id).await,
+            None => FlowNotFoundSnafu { id: flow_id }.fail(),
+        }
+    }
+
+    async fn flow_exist(&self, flow_id: FlowId) -> Result<bool, Error> {
+        let flow_type = self.src_table2flow.read().unwrap().get_flow_type(flow_id);
+        // not using `flow_type.is_some()` to make sure the flow is actually exist in the underlying engine
+        match flow_type {
+            Some(FlowType::Batching) => self.batching_engine.flow_exist(flow_id).await,
+            Some(FlowType::Streaming) => self.streaming_engine.flow_exist(flow_id).await,
+            None => Ok(false),
+        }
+    }
+
+    async fn handle_flow_inserts(
+        &self,
+        request: api::v1::region::InsertRequests,
+    ) -> Result<(), Error> {
+        // TODO(discord9): make as little clone as possible
+        let mut to_stream_engine = Vec::with_capacity(request.requests.len());
+        let mut to_batch_engine = request.requests;
+
+        {
+            let src_table2flow = self.src_table2flow.read().unwrap();
+            to_batch_engine.retain(|req| {
+                let region_id = RegionId::from(req.region_id);
+                let table_id = region_id.table_id();
+                if src_table2flow.in_stream(table_id) {
+                    to_stream_engine.push(req.clone());
+                }
+                if src_table2flow.in_batch(table_id) {
+                    return true;
+                }
```

Comment:
Is it possible to have useless problematic insert requests? If it's an exception, it's recommended to log it.